fem/particleset.hpp

```diff
@@ -0,0 +1,561 @@
+ *  and optionally can have any number of Vector (of arbitrary vdim) and scalar
+ *  integer data in the form of @b fields and @b tags respectively. All particle
+ *  data are internally stored in a Struct-of-Arrays fashion, as elaborated on
+ *  below.
+ *
+ *  @par Coordinates:
+ *  All particle coordinates are stored in a ParticleVector with vector
+ *  dimension equal to the spatial dimension, ordered either byNODES or byVDIM.
+ *
+ *  @par IDs:
+ *  Each particle is assigned a unique global ID of type unsigned long long. In
+ *  parallel, IDs are initialized starting with @b rank and striding by @b size.
+ *
+ *  @par Fields:
+ *  Fields represent scalar or vector \ref real_t data to be associated with
+ *  each particles, such as mass, momentum, or moment. For a given field, all
+ *  particle data is stored in a single ParticleVector with a given
+ *  vector dimension (1 for scalar data) and Ordering::Type (byNODES or
+ *  byVDIM).
+ *
+ *  @par Tags:
+ *  Tags represent an integer to be associated with each particle. For a given
+ *  tag, all particle integers are stored in a single Array<int>.
+ *
+ *  @par Names:
+ *  Each field and tag can optionally be given a name (string) to be used when
+ *  printing particle data in CSV format using PrintCSV().
+ *
+ */
+class ParticleSet
+{
+public:
+   using IDType = unsigned long long;
+private:
+   /// Constructs an Array of size N filled with Ordering::Type o.
+   static Array<Ordering::Type> GetOrderingArray(Ordering::Type o, int N);
+
+   /// Returns default field name for field index i. "Field_{i}"
+   static std::string GetDefaultFieldName(int i);
+
+   /// Returns default tag name for tag index i. "Tag_{i}"
+   static std::string GetDefaultTagName(int i);
+
+   /// Constructs an Array of size N filled with nullptr.
+   static Array<const char*> GetEmptyNameArray(int N);
+
+#ifdef MFEM_USE_MPI
+   static int GetRank(MPI_Comm comm_);
+   static int GetSize(MPI_Comm comm_);
+#endif // MFEM_USE_MPI
+
+protected:
+   /// Stride for IDs (when new particles are added).
+   /** In parallel, this defaults to the number of MPI ranks. */
+   const int id_stride;
+
+   /// Current ID to be assigned to the next particle added.
+   /** In parallel, this starts locally as the rank and increments with
+    * id_stride, ensuring a global unique identifier whenever a particle is
+    * added.
+    */
+   IDType id_counter;
+
+   /// Global unique IDs of particles owned by this rank.
+   Array<IDType> ids;
+
+   /// Spatial coordinates of particles owned by this rank.
+   ParticleVector coords;
+
+   /// All particle fields for particles owned by this rank.
+   std::vector<std::unique_ptr<ParticleVector>> fields;
+
+   /// All particle tags for particles owned by this rank.
+   std::vector<std::unique_ptr<Array<int>>> tags;
+
+   /// Field names, to be written when PrintCSV() is called.
+   std::vector<std::string> field_names;
+
+   /// Tag names, to be written when PrintCSV() is called.
```

Comment:
Does the indexing correspond to the tag's integers. Hence it's expected to have consecutive tags?

Reply:
Each index here is for a given field or tag. For example one may have the first tag represent "Active/Inactive" where maybe an integer 0 is inactive and 1 is active. A second tag may be like "Order" (as in Navier for time-integration), where it is an integer representing what order to use in integration for each particle.

In this case, Tag 0 is the "Active/Inactive" tag, and Tag 1 is the "Order" tag.